File: backend/app/schemas/ai.py
"""Pydantic schemas for the AI interpret endpoint (D4 Phase E).

The wire shape mirrors `docs/design/09-global-ai-input.md` § 9-9 / 9-12 — the
`interpreted` payload is the structured JSON returned by Gemini and almost
always contains an `actions` array (see design 09 § 9-7 for multi-action).

Note: the underlying `ai_interpret_logs` table only has prompt/response/
model/latency_ms/user_id columns. The `context_type` and `cost_usd` fields
needed by D4 Phase E are persisted inside the JSONB `response` column under
the `_meta` wrapper key so we don't need a fresh alembic migration.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

ContextType = Literal[
    "patient_create", "event_create", "override_create", "general"
]


class InterpretRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    prompt: str = Field(min_length=1, max_length=4000, description="自然言語入力")
    context_type: ContextType = Field(default="general")
    context: dict[str, Any] = Field(
        default_factory=dict,
        description=(
            "追加コンテキスト (任意): today, iso_week, weekday, staff_list,"
            " patient_list 等を任意で渡せます。サーバ側でも自動補完します。"
        ),
    )


class InterpretResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    interpreted: dict[str, Any] = Field(
        description="Gemini が返した構造化 JSON。通常は actions[] を含む"
    )
    confidence: float = Field(ge=0.0, le=1.0, description="actions の最高信頼度")
    raw_response: str = Field(description="Gemini の生テキスト (デバッグ用)")
    log_id: UUID = Field(description="AiInterpretLog.id (監査ログ参照用)")
    model: str
    latency_ms: int
    cost_usd: float = Field(ge=0.0)
    context_type: ContextType


class AiLogRead(BaseModel):
    """Enriched read schema that surfaces context_type / cost from the JSONB
    `response._meta` wrapper for the integrations / AI logs page."""

    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: UUID
    prompt: str
    response: dict[str, Any] = Field(default_factory=dict)
    model: str
    latency_ms: int
    user_id: UUID | None = None
    created_at: datetime
    updated_at: datetime
    context_type: str | None = None
    cost_usd: float | None = None
    confidence: float | None = None

File: frontend/lib/schemas/ai.ts
/**
 * AI interpret zod schemas — D4 Phase E / Wave 4-B.
 *
 * Mirrors backend `app/schemas/ai.py`. Used by `useInterpret()` and the
 * `<AiInputModal />` to validate Gemini's structured response before we
 * render it to the operator.
 */
import { z } from 'zod';

// Wave 11 で v2 (10 種類) に揃える。BE `AiContextType` (v2/enums.py) と完全一致。
// 旧 `event_create` / `override_create` は v2 で `staff_event` / `staff_off` に置換。
export const AI_CONTEXT_TYPES = [
  'staff_off',
  'staff_event',
  'staff_mentor',
  'staff_create',
  'patient_create',
  'patient_cancel',
  'patient_reschedule',
  'patient_special_week',
  'staff_status_update',
  'patient_status_update',
  'general',
  'out_of_scope',
] as const;

export const AiContextTypeSchema = z.enum(AI_CONTEXT_TYPES);
export type AiContextType = z.infer<typeof AiContextTypeSchema>;

export const AI_CONTEXT_LABELS: Record<AiContextType, string> = {
  staff_off: 'スタッフ休み (その週だけ)',
  staff_event: 'スタッフイベント (会議・研修)',
  staff_mentor: '新人フラグ + 同行スタッフ割付',
  staff_create: 'スタッフ新規登録',
  patient_create: '患者新規登録',
  patient_cancel: '患者訪問キャンセル',
  patient_reschedule: '患者訪問日時変更',
  patient_special_week: '特別訪問週間 ON/OFF',
  staff_status_update: 'スタッフ状態変更 (在籍/休職/退職)',
  patient_status_update: '患者状態変更 (稼働中/一時休止/入院中/開始前/解約済み)',
  general: '汎用 (自動判定)',
  out_of_scope: '範囲外',
};

export const InterpretRequestSchema = z.object({
  prompt: z.string().min(1, '入力してください').max(4000),
  context_type: AiContextTypeSchema.default('general'),
  context: z.record(z.unknown()).default({}),
});
export type InterpretRequest = z.infer<typeof InterpretRequestSchema>;

/** Single action returned inside `interpreted.actions[]`. The fields object
 * is intentionally left as a free-form record because each action_type has
 * a different shape (see design 09-9 / D4 plan §6). */
export const InterpretedActionSchema = z.object({
  action_type: z.string(),
  confidence: z.number().min(0).max(1).default(0),
  fields: z.record(z.unknown()).default({}),
});
export type InterpretedAction = z.infer<typeof InterpretedActionSchema>;

export const InterpretedPayloadSchema = z
  .object({
    actions: z.array(InterpretedActionSchema).default([]),
  })
  .passthrough();
export type InterpretedPayload = z.infer<typeof InterpretedPayloadSchema>;

export const InterpretResponseSchema = z.object({
  interpreted: InterpretedPayloadSchema,
  confidence: z.number().min(0).max(1),
  raw_response: z.string(),
  log_id: z.string().uuid(),
  model: z.string(),
  latency_ms: z.number().int(),
  cost_usd: z.number(),
  context_type: AiContextTypeSchema,
});
export type InterpretResponse = z.infer<typeof InterpretResponseSchema>;

export const AiLogReadSchema = z.object({
  id: z.string().uuid(),
  prompt: z.string(),
  response: z.record(z.unknown()).default({}),
  model: z.string(),
  latency_ms: z.number().int(),
  user_id: z.string().uuid().nullable().optional(),
  created_at: z.string(),
  updated_at: z.string(),
  context_type: z.string().nullable().optional(),
  cost_usd: z.number().nullable().optional(),
  confidence: z.number().nullable().optional(),
});
export type AiLogRead = z.infer<typeof AiLogReadSchema>;
